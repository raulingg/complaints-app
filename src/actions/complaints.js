import database from '../firebase/firebase';

export const addComplaint = complaint => ({
  type: 'ADD_COMPLAINT',
  complaint,
});

export const startAddComplaint = (complaintData = {}) => (dispatch, getState) => {
  const {
    auth: { uid },
  } = getState();
  const {
    title = '',
    content = '',
    reportedAt = 0,
    happenedAt = 0,
    reportTo = '',
    address = '',
  } = complaintData;
  const complaint = { title, content, reportedAt, happenedAt, reportTo, address, uid };

  return database
    .ref('complaints')
    .push(complaint)
    .then(ref => {
      dispatch(
        addComplaint({
          id: ref.key,
          ...complaint,
        })
      );
    });
};

export const editComplaint = (id, updates) => ({
  type: 'EDIT_COMPLAINT',
  id,
  updates,
});

export const startEditComplaint = (id, updates) => dispatch =>
  database
    .ref(`complaints/${id}`)
    .update(updates)
    .then(() => {
      dispatch(editComplaint(id, updates));
    });

export const setComplaints = complaints => ({
  type: 'SET_COMPLAINTS',
  complaints,
});

export const startSetComplaints = () => dispatch =>
  database
    .ref('complaints')
    .once('value')
    .then(snapshot => {
      const complaints = [];

      snapshot.forEach(childSnapshot => {
        complaints.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });

      dispatch(setComplaints(complaints));
    });
