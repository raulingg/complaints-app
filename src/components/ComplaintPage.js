import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

export const ComplaintPage = ({ complaint }) => (
  <div>
    <h1>{complaint.title}</h1>
    <p>
      Denuncia: <span>{complaint.reportTo}</span>
    </p>
    <div dangerouslySetInnerHTML={{ __html: complaint.content }} />
    <p>
      <span>Denunciado el {moment(complaint.happenedAt).format()}</span>
    </p>
  </div>
);

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

export default connect(mapStateToProps)(ComplaintPage);
