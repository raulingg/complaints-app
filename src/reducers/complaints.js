// Expenses Reducer

const complaintsReducerDefaultState = [];

export default (state = complaintsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_COMPLAINT':
      return [...state, action.complaint];
    case 'REMOVE_COMPLAINT':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_COMPLAINT':
      return state.map(complaint => {
        if (complaint.id === action.id) {
          return {
            ...complaint,
            ...action.updates,
          };
        }
        return complaint;
      });
    case 'SET_COMPLAINTS':
      return action.complaints;
    default:
      return state;
  }
};
