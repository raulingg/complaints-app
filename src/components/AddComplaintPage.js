import React, { Component } from 'react';
import { connect } from 'react-redux';
import ComplaintFormComponent from './ComplaintForm';
import { startAddComplaint as startAddComplaintAction } from '../actions/complaints';

export class AddComplaintPage extends Component {
  onSubmit = complaint => {
    const { history, startAddComplaint } = this.props;
    startAddComplaint(complaint);
    history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Reporta tu denuncia </h1>
        <ComplaintFormComponent onSubmit={this.onSubmit} />
      </div>
    );
  }
}
const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => complaint.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startAddComplaint: complaint => dispatch(startAddComplaintAction(complaint)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AddComplaintPage);
