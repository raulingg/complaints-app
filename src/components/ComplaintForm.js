import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startAddComplaint as startAddComplaintAction } from '../actions/complaints';

export class ComplaintForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.complaint ? props.complaint.title : '',
      content: props.complaint ? props.complaint.content : '',
      reportTo: props.complaint ? props.complaint.reportTo : '',
      happenedAt: props.complaint ? moment(props.complaint.happenedAt) : moment(),
    };
  }

  onTitleChange = e => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  onContentChange = e => {
    const content = e.target.value;
    this.setState(() => ({ content }));
  };

  onReportToChange = e => {
    const reportTo = e.target.value;
    this.setState(() => ({ reportTo }));
  };

  onHappenedAtChange = e => {
    const happenedAt = e.target.value;
    this.setState(() => ({ happenedAt }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { title, content, happenedAt, reportTo } = this.state;
    const { startAddComplaint, history } = this.props;
    startAddComplaint({
      title,
      content,
      happenedAt,
      reportTo,
    });
    history.push('/');
  };

  render() {
    const { title, content, happenedAt, reportTo } = this.state;
    return (
      <form onSubmit={this.onSubmit}>
        <div>
          <input type="text" name="title" value={title} onChange={this.onTitleChange} />
        </div>
        <div>
          <textarea
            name="content"
            maxLength="2000"
            rows="10"
            value={content}
            onChange={this.onContentChange}
          />
        </div>
        <div>
          <input
            type="date"
            name="happenedAt"
            value={happenedAt}
            onChange={this.onHappenedAtChange}
          />
        </div>
        <div>
          <input type="text" name="reportTo" value={reportTo} onChange={this.onReportToChange} />
        </div>
        <div>
          <button className="button">Guardar cambios</button>
        </div>
      </form>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComplaint: complaint => dispatch(startAddComplaintAction(complaint)),
});

export default connect(
  undefined,
  mapDispatchToProps
)(ComplaintForm);
