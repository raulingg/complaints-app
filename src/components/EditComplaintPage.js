import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { compose } from 'recompose';
import ComplaintFormComponent from './ComplaintForm';
import { startEditComplaint as startEditComplaintAction } from '../actions/complaints';
import withAuthorization from './auth/withAuthorization';

export class EditComplaintPage extends Component {
  onSubmit = updates => {
    const {
      history,
      startEditComplaint,
      complaint: { id },
    } = this.props;
    startEditComplaint(id, updates);
    history.push('/');
  };

  render() {
    const { complaint } = this.props;
    return (
      <div>
        <h1>Editar denuncia </h1>
        <ComplaintFormComponent complaint={complaint} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  complaint: state.complaints.find(complaint => String(complaint.id) === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startEditComplaint: (id, complaint) => dispatch(startEditComplaintAction(id, complaint)),
});

export default compose(
  withRouter,
  withAuthorization,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(EditComplaintPage);
