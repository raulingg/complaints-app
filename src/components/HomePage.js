import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import Button from '../styles/components/Buttons';
import { startSetComplaints as startSetComplaintsAction } from '../actions/complaints';
import LoadingPage from './LoadingPage';

export class HomePage extends Component {
  state = {
    isLoading: true,
  };

  async componentDidMount() {
    const { startSetComplaints } = this.props;
    await startSetComplaints();
    this.setState({ isLoading: false });
  }

  render() {
    const { complaints } = this.props;
    const { isLoading } = this.state;

    return isLoading ? (
      <LoadingPage />
    ) : (
      <div>
        <div>
          <div>Denuncia</div>
          <div>Tipo</div>
          <div>Fecha</div>
        </div>
        <div>
          <Button as={Link} to="/add">
            Denunciar
          </Button>
          {complaints.length === 0 ? (
            <div>
              <span>No encontramos ninguna denuncia.</span>
            </div>
          ) : (
            complaints.map(complaint => (
              <div key={complaint.id}>
                <p>
                  <Link to={`show/${complaint.id}`}>{complaint.title}</Link>
                </p>
              </div>
            ))
          )}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  complaints: state.complaints,
});

const mapDispatchToProps = dispatch => ({
  startSetComplaints: () => dispatch(startSetComplaintsAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePage);
