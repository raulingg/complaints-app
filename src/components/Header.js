import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout as startLogoutAction } from '../actions/auth';
import * as HeaderStyles from '../styles/components/Header';
import ContentContainer from '../styles/components/ContentContainer';
import { ButtonLink } from '../styles/components/Buttons';

export const Header = ({ isAuthenticated, startLogout }) => (
  <HeaderStyles.Header>
    <ContentContainer>
      <HeaderStyles.Content>
        <HeaderStyles.Title to="/">
          <h1>Denuncias</h1>
        </HeaderStyles.Title>
        {isAuthenticated ? (
          <ButtonLink onClick={startLogout}>Salir</ButtonLink>
        ) : (
          <ButtonLink as={Link} to="/login">
            Iniciar sesi√≥n
          </ButtonLink>
        )}
      </HeaderStyles.Content>
    </ContentContainer>
  </HeaderStyles.Header>
);

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.user,
});

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogoutAction()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header);
