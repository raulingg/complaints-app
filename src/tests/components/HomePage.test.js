import React from 'react';
import { shallow } from 'enzyme';
import { HomePage } from '../../components/HomePage';
import complaints from '../fixtures/complaints';

describe('HomePage tests', () => {
  const startSetComplaintsSpy = jest.fn(() => Promise.resolve());
  const wrapper = shallow(<HomePage startSetComplaints={startSetComplaintsSpy} complaints={[]} />);
  wrapper.setState({ isLoading: false });

  test('should correctly render ComplaintsList with complaints', () => {
    wrapper.setProps({ complaints });
    expect(wrapper).toMatchSnapshot();
  });

  test('should correctly render ComplaintsList with no complaints', () => {
    wrapper.setState({ isLoading: false });
    wrapper.setProps({ complaints: [] });
    expect(wrapper).toMatchSnapshot();
  });

  test('should correctly render LoadingPage when isLoading is true', () => {
    wrapper.setState({ isLoading: true });
    wrapper.setProps({ complaints: [] });
    expect(wrapper).toMatchSnapshot();
  });
});
