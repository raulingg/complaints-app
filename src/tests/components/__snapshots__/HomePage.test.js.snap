// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage tests should correctly render ComplaintsList with complaints 1`] = `
<div>
  <div>
    <div>
      Denuncia
    </div>
    <div>
      Tipo
    </div>
    <div>
      Fecha
    </div>
  </div>
  <div>
    <Buttons__Button
      as={[Function]}
      to="/add"
    >
      Denunciar
    </Buttons__Button>
    <div
      key="1"
    >
      <p>
        <Link
          replace={false}
          to="show/1"
        >
          Me robaron mi celular
        </Link>
      </p>
    </div>
    <div
      key="2"
    >
      <p>
        <Link
          replace={false}
          to="show/2"
        >
          Me robaron mi celular
        </Link>
      </p>
    </div>
    <div
      key="3"
    >
      <p>
        <Link
          replace={false}
          to="show/3"
        >
          Me robaron mi celular
        </Link>
      </p>
    </div>
  </div>
</div>
`;

exports[`HomePage tests should correctly render ComplaintsList with no complaints 1`] = `
<div>
  <div>
    <div>
      Denuncia
    </div>
    <div>
      Tipo
    </div>
    <div>
      Fecha
    </div>
  </div>
  <div>
    <Buttons__Button
      as={[Function]}
      to="/add"
    >
      Denunciar
    </Buttons__Button>
    <div>
      <span>
        No encontramos ninguna denuncia.
      </span>
    </div>
  </div>
</div>
`;

exports[`HomePage tests should correctly render LoadingPage when isLoading is true 1`] = `<LoadingPage />`;
